quinn-proto/src/cid_state.rs

```diff
@@ -0,0 +1,196 @@
+            }
+        }
+
+        // Check if retirement of all CIDs that reach their lifetime is still needed
+        // If yes (return true), a new CID must be pushed with updated `retire_prior_to` field to remote peer.
+        // If no (return false), it means remote peer has proactively retired those CIDs (for other reasons) before CID lifetime is reached.
+        (current_retire_prior_to..self.retire_seq).any(|seq| self.active_seq.contains(&seq))
+    }
+
+    /// Update cid state when `NewIdentifiers` event is received
+    pub(crate) fn new_cids(&mut self, ids: &[IssuedCid], now: Instant) {
+        // `ids` could be `None` once active_connection_id_limit is set to 1 by peer
+        if let Some(last_cid) = ids.last() {
+            self.issued += ids.len() as u64;
+            // Record the timestamp of CID with the largest seq number
+            let sequence = last_cid.sequence;
+            ids.iter().for_each(|frame| {
+                self.active_seq.insert(frame.sequence);
+            });
+            self.track_lifetime(sequence, now);
+        }
+    }
+
+    /// Update CidState when recieve a `RETIRE_CONNECTION_ID` frame
+    /// Return a boolean variable or `TransportError` if CID content violates RFC
+    /// When boolean variable is `true`, a new CID should be issued to peer.
+    pub(crate) fn on_cid_retirement(
+        &mut self,
+        sequence: u64,
+        limit: u64,
+    ) -> Result<bool, TransportError> {
+        if self.cid_len == 0 {
+            return Err(TransportError::PROTOCOL_VIOLATION(
+                "RETIRE_CONNECTION_ID when CIDs aren't in use",
+            ));
+        }
+        if sequence > self.issued {
+            debug!(
+                sequence,
+                "got RETIRE_CONNECTION_ID for unissued sequence number"
+            );
+            return Err(TransportError::PROTOCOL_VIOLATION(
+                "RETIRE_CONNECTION_ID for unissued sequence number",
+            ));
+        }
+        self.active_seq.remove(&sequence);
+        // Consider a scenario where peer A has active remote cid 0,1,2.
+        // Peer B first send a NEW_CONNECTION_ID with cid 3 and retire_prior_to set to 1.
+        // Peer A processes this NEW_CONNECTION_ID frame; update remote cid to 1,2,3
+        // and meanwhile send a RETIRE_CONNECTION_ID to retire cid 0 to peer B.
+        // If peer B doesn't check the cid limit here and send a new cid again, peer A will then face CONNECTION_ID_LIMIT_ERROR
+        let allow_more_cids = limit > self.active_seq.len() as u64;
+
+        Ok(allow_more_cids)
+    }
+
+    /// Length of local Connection IDs
+    pub(crate) fn cid_len(&self) -> usize {
+        self.cid_len
+    }
+
+    /// The value for `retire_prior_to` field in `NEW_CONNECTION_ID` frame
+    pub(crate) fn retire_prior_to(&self) -> u64 {
+        self.retire_seq
+    }
+
+    #[cfg(test)]
+    pub(crate) fn active_seq(&self) -> (u64, u64) {
```

Comment:
Maybe this should return a `Range`? Also iterating the same set twice seems a little ugly/inefficient, maybe just write an old-fashioned for-loop that tracks both minimum and maximum?

Reply:
Ah i see. Make sense

Reply:
Essentially, we don't know the number in this hashset is continuous or not (so `Range` probably is not a good option here), and we only need the max and min from the set.